generator/dart/generated/google_cloud_protobuf/lib/src/protobuf.p.dart

```diff
@@ -14,21 +14,106 @@
 
 part of '../protobuf.dart';
 
-class FieldMaskHelper {
-  /// Encode the field mask as a single comma-separated string.
-  static String encode(FieldMask fieldMask) {
-    return fieldMask.paths?.join(',') ?? '';
+/// `Any` contains an arbitrary serialized message along with a URL that
+/// describes the type of the serialized message.
+class Any extends Message {
+  static const Set<String> _customEncodedTypes = {
+    'google.protobuf.Duration',
+    'google.protobuf.FieldMask'
+  };
+
+  /// The raw JSON encoding of the underlying value.
+  final Map<String, dynamic> json;
+
+  Any({Map<String, dynamic>? json}) : this.json = json ?? {};
+
+  /// Create an [Any] from an existing [message].
+  ///
+  /// [fullyQualifiedName] should be the fully qualified type name of the given
+  /// message (for example, `google.protobuf.Duration`).
+  Any.from(Message message, String fullyQualifiedName) : json = {} {
+    packInto(message, fullyQualifiedName);
   }
 
-  /// Decode the field mask from a single comma-separated string.
-  static FieldMask decode(String format) {
-    return FieldMask(paths: format.split(','));
+  factory Any.fromJson(Map<String, dynamic> json) {
+    return Any(json: json);
+  }
+
+  /// '@type' will be something like
+  /// `type.googleapis.com/google.protobuf.Duration`, or
+  /// `type.googleapis.com/google.rpc.ErrorInfo`.
+  String get _type => json['@type'];
+
+  /// Return the fully qualified name of the contained type.
+  ///
+  /// For example, `google.protobuf.Duration`, or `google.rpc.ErrorInfo`.
+  String get typeName {
+    final type = _type;
+
+    final index = type.lastIndexOf('/');
+    if (index == -1) return type;
+
+    return type.substring(index + 1);
   }
+
+  /// Returns whether the type represented by this `Any` is the same as [name].
+  ///
+  /// [name] should be a fully qualified type name, for example,
+  /// `google.protobuf.Duration` or `google.rpc.ErrorInfo`.
+  bool isType(String name) => typeName == name;
+
+  /// Deserialize a message from this `Any` object.
+  ///
+  /// For most message types, you should pass the `<type>.fromJson` constructor
+  /// into this method. Eg.:
+  ///
+  /// ```dart
+  /// if (any.isType(Status.fullyQualifiedName)) {
+  ///   final status = any.unpackFrom(Status.fromJson);
+  ///   ...
+  /// }
+  /// ```
+  T unpackFrom<T, S>(T Function(S) decoder) {
+    final name = typeName;
+
+    if (_customEncodedTypes.contains(name)) {
+      // Handle custom types:
+      //   { "@type": "type.googl...obuf.Duration", "value": "1.212s" }
+      return decoder(json['value'] as S);
+    } else {
+      return decoder(json as S);
+    }
+  }
+
+  /// Serialize the given message into this `Any` instance.
+  ///
+  /// [fullyQualifiedName] should be the fully qualified type name of the given
+  /// message (for example, `google.protobuf.Duration`).
+  void packInto(Message message, String fullyQualifiedName) {
```

Comment:
Here, we require the user to pass in the FQN name of the given message. We could generate an instance member for `Message` which would respond w/ the fully qualified type name. However, that would add a new member into the namespace of every message that would really only be used for `Any.packInto` - I'm not sure its worth it for this single use-case.

Reply:
It seems sad that we are asking folks to provide a value (the fully qualified name) when we know what the answer should be.  We should be able to prevent bugs like this one:

```dart
  any.packInto(Duration(...), "type.googleapis.com/google.protobuf.Timestamp")
```

----

Could we make this into a generic function and access the type name through the class?

```dart
void packInto<M extends Message>(M message) {
  // Can we use `M.fullyQualifiedName` here ??  Or maybe some helper generic `SomeTrait<M>().fullyQualifiedName()` ?
}
```